src/equations/serre_green_naghdi_1d.jl

```diff
@@ -0,0 +1,960 @@
+        if equations.bathymetry isa BathymetryMildSlope
+            factor = 0.75
+        else # equations.bathymetry isa BathymetryVariable
+            factor = 1.0
+        end
+        # The code below is equivalent to
+        #   dv .= (Diagonal(h .+ 0.75 .* h .* b_x.^2) - Dmat * (Diagonal(1/3 .* h.^3) * Dmat - Diagonal(0.5 .* h.^2 .* b_x)) - Diagonal(0.5 .* h.^2 .* b_x) * Dmat) \ tmp
+        # but faster since the symbolic factorization is reused.
+        # Floating point errors accumulate a bit and the system matrix
+        # is not necessarily perfectly symmetric but only up to round-off
+        # errors. We wrap it here to avoid issues with the factorization.
+        @. M_h_p_h_bx2 = h + factor * h * b_x^2
+        scale_by_mass_matrix!(M_h_p_h_bx2, D)
+        inv3 = 1 / 3
+        @. M_h3_3 = inv3 * h^3
+        scale_by_mass_matrix!(M_h3_3, D)
+        @. M_h2_bx = 0.5 * h^2 * b_x
+        scale_by_mass_matrix!(M_h2_bx, D)
+        system_matrix = Symmetric(Diagonal(M_h_p_h_bx2)
+                                  +
+                                  Dmat_minus' * (Diagonal(M_h3_3) * Dmat_minus
+                                   -
+                                   Diagonal(M_h2_bx))
+                                  -
+                                  Diagonal(M_h2_bx) * Dmat_minus)
+    end
+
+    @trixi_timeit timer() "solving elliptic system" begin
+        if issparse(system_matrix)
+            (; factorization) = cache
+            cholesky!(factorization, system_matrix; check = false)
+            if issuccess(factorization)
+                scale_by_mass_matrix!(tmp, D)
+                dv .= factorization \ tmp
+            else
+                # The factorization may fail if the time step is too large
+                # and h becomes negative.
+                fill!(dv, NaN)
+            end
+        else
+            factorization = cholesky!(system_matrix)
+            scale_by_mass_matrix!(tmp, D)
+            ldiv!(dv, factorization, tmp)
+        end
+    end
+
+    return nothing
+end
+
+@inline function prim2cons(q, equations::SerreGreenNaghdiEquations1D)
+    h = waterheight(q, equations)
+    v = velocity(q, equations)
+    b = bathymetry(q, equations)
+
+    hv = h * v
+    return SVector(h, hv, b)
+end
+
+@inline function cons2prim(u, equations::SerreGreenNaghdiEquations1D)
+    h, hv, b = u
+
+    eta = h + b
+    v = hv / h
+    D = equations.eta0 - b
+    return SVector(eta, v, D)
+end
+
+@inline function waterheight_total(q, equations::SerreGreenNaghdiEquations1D)
+    return q[1]
+end
+
+@inline function velocity(q, equations::SerreGreenNaghdiEquations1D)
+    return q[2]
+end
+
+@inline function bathymetry(q, equations::SerreGreenNaghdiEquations1D)
+    D = q[3]
+    return equations.eta0 - D
+end
+
+@inline function waterheight(q, equations::SerreGreenNaghdiEquations1D)
+    return waterheight_total(q, equations) - bathymetry(q, equations)
+end
```

Comment:
This function does not depend on the concrete equation, right? So could we put it in `src/equations/equations.jl` and dispatch on `AbstractShallowWaterEquations` (and remove it from the other files, too)?

Reply:
Yes, we could - but it assumes that we use the variables `eta, v, D` since we would have more efficient ways to compute it otherwise. However, since you wanted to use these variables for consistency (which makes this PR so annoying to review), I will move it 👍